src/SWP/Bundle/CoreBundle/Theme/Processor/RequiredDataProcessor.php

```diff
@@ -0,0 +1,79 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Superdesk Web Publisher Core Bundle.
+ *
+ * Copyright 2017 Sourcefabric z.ú. and contributors.
+ *
+ * For the full copyright and license information, please see the
+ * AUTHORS and LICENSE files distributed with this source code.
+ *
+ * @copyright 2017 Sourcefabric z.ú
+ * @license http://www.superdesk.org/license
+ */
+
+namespace SWP\Bundle\CoreBundle\Theme\Processor;
+
+use SWP\Bundle\ContentBundle\Service\RouteServiceInterface;
+use SWP\Bundle\CoreBundle\Theme\Generator\GeneratorInterface;
+use SWP\Bundle\CoreBundle\Theme\Model\ThemeInterface;
+
+class RequiredDataProcessor implements RequiredDataProcessorInterface
+{
+    /**
+     * @var RouteServiceInterface
+     */
+    protected $themeRoutesGenerator;
+
+    /**
+     * @var GeneratorInterface
+     */
+    protected $themeMenusGenerator;
+
+    /**
+     * @var GeneratorInterface
+     */
+    protected $themeContainersGenerator;
+
+    /**
+     * @var GeneratorInterface
+     */
+    protected $themeWidgetsGenerator;
+
+    /**
+     * @var GeneratorInterface
+     */
+    protected $themeContentListsGenerator;
+
+    /**
+     * RequiredDataProcessor constructor.
+     *
+     * @param GeneratorInterface $themeRoutesGenerator
+     * @param GeneratorInterface $themeMenusGenerator
+     * @param GeneratorInterface $themeContainersGenerator
+     * @param GeneratorInterface $themeWidgetsGenerator
+     * @param GeneratorInterface $themeContentListsGenerator
+     */
+    public function __construct(GeneratorInterface $themeRoutesGenerator, GeneratorInterface $themeMenusGenerator, GeneratorInterface $themeContainersGenerator, GeneratorInterface $themeWidgetsGenerator, GeneratorInterface $themeContentListsGenerator)
+    {
+        $this->themeRoutesGenerator = $themeRoutesGenerator;
+        $this->themeMenusGenerator = $themeMenusGenerator;
+        $this->themeContainersGenerator = $themeContainersGenerator;
+        $this->themeWidgetsGenerator = $themeWidgetsGenerator;
+        $this->themeContentListsGenerator = $themeContentListsGenerator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function processTheme(ThemeInterface $theme): void
+    {
+        $this->themeRoutesGenerator->generate($theme->getRoutes());
```

Comment:
What about creating a "Chain Generator":
- create a compiler pass and register existing generators using custom tag (add generators to the `ThemeGeneratorClass`) 
- Create `ThemeGeneratorChain` class which will have `supports` and `generate` method where you will loop foreach generator
- add method `supports` to each generator so it supports only specific object types

It will eliminate coupling a lot and it will be easier to add new generators later by just registering a new generator with a custom tag in config, without a need to change that class and inject more dependencies to the construct.

Reply:
We can think about this later if we will add new things here. For now i don't see issues with this implementation.